### PR TITLE
ref(typing): Few typing changes for sentry.snuba.metrics

### PR DESCRIPTION
**Pull request by armenzg**

### CODE REVIEW

File: src/sentry/snuba/metrics/query.py [hunk 1]
```diff
@@ -243,7 +243,7 @@ def validate_orderby(self) -> None:
                 self._validate_field(metric_order_by_field.field)
 
         orderby_metric_fields: Set[MetricField] = set()
-        metric_entities: Set[MetricField] = set()
+        metric_entities: Set[MetricEntity] = set()
```
Comment (armenzg) [inline src/sentry/snuba/metrics/query.py:246]: @obostjancic @iambriccardo Could you please check this type change?

**Review by asottile-sentry — APPROVED** [review]

File: src/sentry/snuba/metrics/query.py [hunk 2]
```diff
@@ -290,6 +290,8 @@ def action_by_str_fields(self, on_group_by: bool) -> Set[str]:
     def validate_limit(self) -> None:
         if self.limit is None:
             return
+        if self.start is None or self.end is None:
+            raise Exception("We need start and end to be set.")
```
Comment (armenzg) [inline src/sentry/snuba/metrics/query.py:294]: `start` and `end` are marked as `Optional[datetime]` but the function below requires both values.

File: src/sentry/snuba/metrics/utils.py [hunk 3]
```diff
@@ -430,7 +430,7 @@ def to_intervals(
 
 def get_num_intervals(
     start: Optional[datetime],
-    end: datetime,
+    end: Optional[datetime],
```
Comment (armenzg) [inline src/sentry/snuba/metrics/utils.py:433]: As you can see the function this function calls has it as `optional`: https://github.com/getsentry/sentry/blob/f2d0a2852daa363da66902fb8ff668433bafcb22/src/sentry/snuba/metrics/utils.py#L393-L395

File: src/sentry/snuba/metrics/query.py (hunk 1 above)
Comment (obostjancic) [reply src/sentry/snuba/metrics/query.py:246]: 👍 Good catch!

**Review by iambriccardo — APPROVED** [review]

**Pull request merged**